sys/log/full/src/log_fcb_bmark.c

```diff
@@ -61,74 +132,258 @@ log_fcb_rotate_bmarks(struct fcb_log *fcb_log)
         bmark = &fcb_log->fl_bset.lfs_bmarks[i];
+#if MYNEWT_VAL(LOG_FCB)
+        if (!fcb_log->fl_bset.lfs_bmarks[i].lfb_entry.fe_area) {
+#if MYNEWT_VAL(LOG_FCB_SECTOR_BOOKMARKS)
+            if (i < fcb_log->fl_fcb.f_sector_cnt) {
+                /* Jump to the non-sector bookmarks since sector
+                 * bookmarks are empty here on
+                 */
+                i = fcb_log->fl_fcb.f_sector_cnt - 1;
+                continue;
+            }
+#else
+            /* Empty non-sector bookmark, nothing more to do
+             * Previous closest bookmark is the closest one */
+            break;
+#endif
+        }
+#elif MYNEWT_VAL(LOG_FCB2)
+        if (!fcb_log->fl_bset.lfs_bmarks[i].lfb_entry.fe_range) {
+#if MYNEWT_VAL(LOG_FCB_SECTOR_BOOKMARKS)
+            if (i < fcb_log->fl_fcb.f_sector_cnt) {
+                /* Jump to the non-sector bookmarks since sector
+                 * bookmarks are empty here on
+                 */
+                i = fcb_log->fl_fcb.f_sector_cnt - 1;
+                continue;
+            }
+#else
+            /* Empty non-sector bookmark, nothing more to do
+             * Previous closest bookmark is the closest one */
+            break;
+#endif
+        }
+#endif
         if (bmark->lfb_index <= index) {
             diff = index - bmark->lfb_index;
-            if (diff < min_diff) {
-                min_diff = diff;
+            if (diff < *min_diff) {
+                *min_diff = diff;
                 closest = bmark;
+                MODLOG_DEBUG(LOG_MODULE_DEFAULT, "index: %u, closest bmark idx: %u, \n",
+                             (unsigned int)index,
+                             (unsigned int)bmark->lfb_index);
+                /* We found the exact match, no need to keep searching for a
+                 * better match
+                 */
+                if (*min_diff == 0) {
+                    break;
+                }
             }
         }
     }
 
     return closest;
 }
 
+#if MYNEWT_VAL(LOG_FCB_SECTOR_BOOKMARKS)
 #if MYNEWT_VAL(LOG_FCB)
-void
-log_fcb_add_bmark(struct fcb_log *fcb_log, const struct fcb_entry *entry,
-                  uint32_t index)
+static void
+log_fcb_insert_sect_bmark(struct fcb_log *fcb_log, struct fcb_entry *entry,
+                          uint32_t index, int pos)
 #elif MYNEWT_VAL(LOG_FCB2)
-void
-log_fcb_add_bmark(struct fcb_log *fcb_log, const struct fcb2_entry *entry,
-                  uint32_t index)
+static void
+log_fcb_insert_sect_bmark(struct fcb_log *fcb_log, struct fcb2_entry *entry,
+                          uint32_t index, int pos)
 #endif
 {
     struct log_fcb_bset *bset;
 
     bset = &fcb_log->fl_bset;
 
-    if (bset->lfs_cap == 0) {
-        return;
+    if (bset->lfs_size && bset->lfs_size > pos && pos != (bset->lfs_cap - 1)) {
+        memmove(&bset->lfs_bmarks[pos + 1], &bset->lfs_bmarks[pos],
+                sizeof(bset->lfs_bmarks[0]) * (bset->lfs_size - pos));
```

Comment:
This code keeps sector bookmarks sorted by index.
It does not look like any other place in the code takes advantage of this order. When bookmarks are search for match this order is not utilized.
Maybe there is not need to move chunk of memory every time when sector bookmark is inserted which seems to be the case since sector are scanned from the oldest one. So every sector bookmark result in memmove with increasing size.

Reply:
This was done because earlier absolute bookmarks could actually be in between the other bookmarks based on the index. Even now, insertion happens as pos 0 but, we could optimize it so that there is no memmove happening. Let me try it.

Reply:
Ok, I did optimize the code quite a bit. Please take a look. As per the review comment I have removed memmove.